### ISSUE TITLE
RO-Crate context should use https for schema.org IRIs

### ISSUE THREAD

**Issue by dnlbauer** [ResearchObject/ro-crate#349]
In recent years, the use of HTTPS over HTTP to reference resources on the web has become the de facto standard across the internet. Today, many modern browsers have started to automatically redirect, or in some cases, even block traffic that attempts to use HTTP instead of the more secure HTTPS. This transition is generally seamless, as redirecting HTTP requests to HTTPS has become a common practice among HTTP and proxy servers.

Unfortunately, for linked data, using HTTP for vocabularies is still common. This is also true for RO-Crates: While the specification provides the jsonld context for RO-Crates via https (https://www.researchobject.org/ro-crate/specification/1.1/context.jsonld), internally the context is not consequent in what scheme it uses. For schema.org, the vocabulary uses HTTP (http://schema.org). For Bioschemas it uses HTTPS https://bioschemas.org.

Schema.org on the other hand, completed it's movement to HTTPS with [Release 12.0](https://webschemas.org/docs/releases.html#v12.0) already in 2021:

> - [PR #2814](https://github.com/schemaorg/schemaorg/pull/2814): Completed process of moving Schema.org to https. This step included the move to https of vocabulary term definitions and consequent change to https of the canonical URI displayed under the term pages [more...] tag. The web site will continue to respond to both http and https URLs. Download files will continue to support both protocols.

As part of this move, the context for schema.org is now available for HTTPS and HTTP separately, but HTTPS is the preferred version.

I suggest to also move all term definitions in the RO-Crate context to use HTTPS for schema.org. Not only is https more secure, but it would also provide consistency in the used scheme. This makes it easier to work with the specification, especially for less experienced users.

**Comment by dnlbauer** [comment]
## Why mixing schemes can be problematic

Algorithms working with JSON-LD usually treat term definitions and the IRIs they expand to as string literals. For example, the IRIs [https://schema.org/author](https://schema.org/author) and [http://schema.org/author](http://schema.org/author)  convey the same meaning to a human reader (and they in fact point to the same resource due to HTTP redirection), but software operating on these terms treat them as being different strings and thus different terms.

This can quickly lead to bugs in cases where multiple contexts are mixed that do not use schemes uniformly.
### Example 

The following excerpt from an RO-Crate defines a dataset with an author. As suggested by the specification, it uses the RO-Crate context with HTTPS. Additionally, it uses type coercion to allow to reference the authors `@id` directly as string, without using a construct such as `{"@id": "https://orcid.org/0000-0001-9447-460X"}`.

```json
{
  "@context": [
    "https://w3id.org/ro/crate/1.1/context",
    {
      "author": {
        "@id": "https://schema.org/author",
        "@type": "@id"
      }
    },
  ],
  "@graph": [
    {
      "@id": "./",
      "@type": "Dataset",
      "name": "Test",
      "author": "https://orcid.org/0000-0001-9447-460X"
    },
    {
      "@id": "https://orcid.org/0000-0001-9447-460X",
      "@type": "Person",
      "name": "Daniel Bauer"
    }
  ]
}
```

Without looking into the RO-Crate context, it's reasonable to use https://schema.org/author here, isn't it? After all, the RO-Crate context is also using HTTPS, so what could possibly go wrong.

Well, expanding this document (i.e. with the library pyld), leads to the document shown below. All terms are now represented with HTTP-based IRIs, except for the author, which now uses HTTPS.

```json
[
  {
    "@id": "./",
    "@type": ["http://schema.org/Dataset"],
    "https://schema.org/author": [
      {"@id": "https://orcid.org/0000-0001-9447-460X"}
    ],
    "http://schema.org/name": [
      {"@value": "Test"}
    ]
  },
  {
    "@id": "https://orcid.org/0000-0001-9447-460X",
    "@type": ["http://schema.org/Person"],
    "http://schema.org/name": [
      {"@value": "Daniel Bauer"}
    ]
  }
]
```

If we flatten this graph to a different context, i.e. only using the RO-Crate context to eliminate the type coercion, not all terms are flattened to the expected context due to this scheme mismatch:

```json
{
  "@context": "https://w3id.org/ro/crate/1.1/context",
  "@graph": [
    {
      "@id": "./",
      "@type": "Dataset",
      "name": "Test"
      "https://schema.org/author": { "@id": "https://orcid.org/0000-0001-9447-460X" }
    },
    {
      "@id": "https://orcid.org/0000-0001-9447-460X",
      "@type": "Person",
      "name": "Daniel Bauer"
    }
  ]
}
```

While this could easily be resolved by using the HTTP form of schema.org for the type coercion of the author, I feel that errors like this can arise too easily with the current RO-Crate specification that suggests HTTPS for the context, but uses HTTP internally.

Especially for the inexperienced user, it's very hard to spot this subtle difference. By being more consistent (=using HTTPS everywhere), we can circumvent such problems.

**Comment by stain** [comment]
I'm afraid schema.org in their JSON-LD context also use `http` variant (even if you request it as `https`), and we most want to be compatible with that.

```
(base) stain@xenamint:~/src/ro-crate$ curl -sI https://schema.org/ | grep -i Link:
link: </docs/jsonldcontext.jsonld>; rel="alternate"; type="application/ld+json"
```

```
(base) stain@xenamint:~/src/ro-crate$ curl -s https://schema.org/docs/jsonldcontext.jsonld | grep schema.org | head
        "@vocab": "http://schema.org/",
        "schema": "http://schema.org/",
        "organizer": { "@id": "schema:organizer"},
```

![image](https://github.com/user-attachments/assets/6213eccb-aa5d-4f75-bd71-52d7cf81fe7b)

You are right that there was a schema.org release with `https` mapping inside context, but it was reverted thereafter, strangely I can't find it now, but it was around version 11-13. Schema.org continue to provide both `http` and https` variants, see https://schema.org/docs/developers.html. This does however mean an incompatibility with RDFa embedded in HTML where `https` variants are used in current examples, as well as with browsers which redirect to `https` version. 

However as RO-Crate uses JSON-LD we want primarily to be compatible with typical usage of schema.org in JSON-LD markup (e.g. as shown from N-Quads output of https://json-ld.org/playground/). 

It is a long-standing principle in RDF and Linked Data world to not rename namespaces unless governance has changed. 

What can be a source of confusion is that schema.org in their wisdom use `"@context": "https://schema.org/` (or sometimes even without `/` or `https://` to be non-URI) -- however in JSON-LD this is **not** a namespace, but an instruction for where to load the JSON-LD context from, which rightly should be done over HTTPS for secure transport.

Note that RO-Crate's JSON-LD context is retrieved over `https` fully with [redirect from https://w3id.org/ro/crate/1.1/context](https://github.com/perma-id/w3id.org/blob/master/ro/crate/.htaccess#L8) to https://www.researchobject.org/ro-crate/specification/1.1/context.jsonld

**Comment by stain** [comment]
Note that we have tried in our specification to be consistent also to link to `http` terms - e.g. https://www.researchobject.org/ro-crate/specification/1.1/root-data-entity.html describes:

_The Root Data Entity is a [Dataset](http://schema.org/Dataset) that represent the RO-Crate as a whole;_  where the link goes to `http://schema.org/Dataset`

**Comment by stain** [comment]
See https://github.com/ResearchObject/ro-crate/blob/main/scripts/schema-context.py for how our context file is generated, I don't know why you would need to re-declare `author` if it is already in the base context. The `"@type": "@id"` pattern is not currently recommended as it can break parsers who treat RO-Crate as JSON without JSON-LD processing - they would see plain strings rather than references to contextual entities.

**Comment by dnlbauer** [comment]
Thanks for the additional context on the topic @stain. I understand the reasoning.

Regarding the question why you would need to re-declare `author`: For the given case, we had the `author` and other data stored as separate digital objects. All objects were valid JSON-LD but we used `https://schema.org` internally.
The problem described here arose when we merged the JSON-LD objects into a single graph to create the RO-Crate. I used [pyld](https://github.com/digitalbazaar/pyld) for this and noticed that the context for `author` and other fields were not flattened correctly into the RO-Crate context due to the mismatch between the schemes.

I still think Linked Data should make the move to `https` (or find othre ways on how to deal with different schemes that have the same semantic meaning). Nevertheless, I agree that for RO-Crate its more useful to use what has the best support.
